File: inkwire/receivers/python-fastapi/inkwire_receiver/handle.py
from datetime import datetime, timezone
from pydantic import ValidationError
from .schema import PostInput
from .auth import authorize
from .render import render_markdown, slugify
from .errors import InkwireError
HEADERS = {"Inkwire-Version": "1"}

async def _dedupe_slug(store, key, base):
    slug, n = base, 1
    while await store.find_by_slug(key, slug):
        n += 1; slug = f"{base}-{n}"
    return slug

async def handle_post(auth_header, raw_body, api_keys, store):
    try:
        key = authorize(auth_header, api_keys)
        try:
            p = PostInput(**(raw_body or {}))
            p.validate_slug()
        except (ValidationError, ValueError) as e:
            raise InkwireError("invalid_payload", str(e).splitlines()[0])
        if p.slug:
            owner = await store.find_by_slug(key, p.slug)
            if owner and owner["external_id"] != p.external_id:
                raise InkwireError("conflict", f"slug '{p.slug}' is taken")
            slug = p.slug
        else:
            existing = await store.find_by_external_id(key, p.external_id)
            slug = existing["slug"] if existing else await _dedupe_slug(store, key, slugify(p.title))
        html = render_markdown(p.markdown)
        result = await store.upsert(key, {
            "external_id": p.external_id, "title": p.title, "html": html, "markdown": p.markdown,
            "slug": slug, "status": p.status,
            "published_at": p.published_at or datetime.now(timezone.utc).isoformat(),
        })
        return 200, {"id": result["id"], "external_id": p.external_id, "url": result["url"],
                     "slug": result["slug"], "status": p.status, "created": result["created"]}, HEADERS
    except InkwireError as e:
        return e.http_status, {"error": {"code": e.code, "message": e.message}}, HEADERS
    except Exception:
        return 500, {"error": {"code": "internal", "message": "unexpected error"}}, HEADERS

File: inkwire/receivers/core-ts/src/handle.ts
import { PostInput } from "./schema.js";
import { authorize } from "./auth.js";
import { renderMarkdown, slugify } from "./render.js";
import { InkwireError } from "./errors.js";
import type { PostStore } from "./store.js";
export type { PostStore } from "./store.js";

const HEADERS = { "Inkwire-Version": "1", "content-type": "application/json" };

export async function handlePost(input: {
  authHeader?: string; rawBody: unknown; apiKeys: string[]; store: PostStore;
}): Promise<{ status: number; body: object; headers: Record<string, string> }> {
  try {
    const key = authorize(input.authHeader, input.apiKeys);
    const parsed = PostInput.safeParse(input.rawBody);
    if (!parsed.success) {
      const first = parsed.error.issues[0];
      throw new InkwireError("invalid_payload", `${first.path.join(".") || "body"}: ${first.message}`);
    }
    const p = parsed.data;

    // Slug resolution + explicit-collision guard.
    let slug: string;
    if (p.slug) {
      const owner = await input.store.findBySlug(key, p.slug);
      if (owner && owner.external_id !== p.external_id) throw new InkwireError("conflict", `slug '${p.slug}' is taken`);
      slug = p.slug;
    } else {
      const existing = await input.store.findByExternalId(key, p.external_id);
      slug = existing?.slug ?? await dedupeSlug(input.store, key, slugify(p.title));
    }

    const html = renderMarkdown(p.markdown);
    const seo = {
      title: p.seo?.title ?? p.title,
      description: p.seo?.description ?? (p.excerpt || p.title),
      image_url: p.seo?.image_url ?? p.cover_image_url,
      noindex: p.seo?.noindex ?? false,
    };
    const result = await input.store.upsert(key, {
      external_id: p.external_id, title: p.title, html, markdown: p.markdown, slug,
      excerpt: p.excerpt, tags: p.tags, cover_image_url: p.cover_image_url,
      canonical_url: p.canonical_url, seo, author: p.author, status: p.status,
      published_at: p.published_at ?? new Date().toISOString(),
    });
    return { status: 200, headers: HEADERS, body: {
      id: result.id, external_id: p.external_id, url: result.url,
      slug: result.slug, status: p.status, created: result.created,
    } };
  } catch (err) {
    const e = err instanceof InkwireError ? err : new InkwireError("internal", "unexpected error");
    return { status: e.httpStatus, headers: HEADERS, body: { error: { code: e.code, message: e.message } } };
  }
}

const MAX_SLUG_LENGTH = 255;

async function dedupeSlug(store: PostStore, key: string, base: string): Promise<string> {
  let slug = base, n = 1;
  while (await store.findBySlug(key, slug)) {
    const suffix = `-${++n}`;
    slug = base.slice(0, MAX_SLUG_LENGTH - suffix.length) + suffix;
  }
  return slug;
}
